crates/bitwarden/src/tool/generators/password.rs

```diff
@@ -40,10 +41,249 @@ pub struct PassphraseGeneratorRequest {
+}
+
+pub(super) fn password_with_rng(
+    mut rng: impl RngCore,
+    input: PasswordGeneratorRequest,
+) -> Result<String> {
+    // We always have to have at least one character set enabled
+    if !input.lowercase && !input.uppercase && !input.numbers && !input.special {
+        return Err(Error::Internal(
+            "At least one character set must be enabled",
+        ));
+    }
+
+    // Generate all character dictionaries
+    let chars = PasswordGeneratorCharSet::new(
+        input.lowercase,
+        input.uppercase,
+        input.numbers,
+        input.special,
+        input.avoid_ambiguous.unwrap_or(false),
+    );
+
+    // Make sure the minimum values are zero when the character
+    // set is disabled, and at least one when it's enabled
+    fn get_minimum(min: Option<u8>, enabled: bool) -> u8 {
+        if enabled {
+            u8::max(min.unwrap_or(1), 1)
+        } else {
+            0
+        }
+    }
+    let min_lowercase = get_minimum(input.min_lowercase, input.lowercase);
+    let min_uppercase = get_minimum(input.min_uppercase, input.uppercase);
+    let min_number = get_minimum(input.min_number, input.numbers);
+    let min_special = get_minimum(input.min_special, input.special);
+
+    // Check that the minimum lengths aren't larger than the password length
+    let min_length = min_lowercase + min_uppercase + min_number + min_special;
+    let length = input.length.unwrap_or(DEFAULT_PASSWORD_LENGTH);
+    if min_length > length {
+        return Err(Error::Internal(
+            "Password length can't be less than the sum of the minimums",
+        ));
+    }
+
+    // Generate the minimum chars of each type, then generate the rest to fill the expected length
+    let mut buf = Vec::with_capacity(length as usize);
+
+    for _ in 0..min_lowercase {
+        buf.push(*chars.lower.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_uppercase {
+        buf.push(*chars.upper.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_number {
+        buf.push(*chars.number.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_special {
+        buf.push(*chars.special.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in min_length..length {
+        buf.push(*chars.all.choose(&mut rng).expect("slice is not empty"));
+    }
+
+    buf.shuffle(&mut rng);
+    Ok(buf.iter().collect())
 }
 
 pub(super) fn passphrase(_input: PassphraseGeneratorRequest) -> Result<String> {
     Ok("correct-horse-battery-staple".to_string())
 }
+
+#[cfg(test)]
+mod test {
+    use rand::SeedableRng;
+
+    use super::*;
+
+    // We convert the slices to Strings to be able to use `contains`
+    // This wouldn't work if the character sets were ordered differently, but that's not the case for us
+    fn to_string(chars: &[char]) -> String {
+        chars.iter().collect()
+    }
+
+    #[test]
+    fn test_password_characters_all() {
```

Comment:
💭 My gut says the unit test names should be more descriptive, but I have no idea how to make that compatible w/ rust's idioms.

Reply:
I haven't seen any recommended practices about testing naming conventions, but beyond splitting the tests into more functions with more descriptive names, the assert macros also have messages that they can print when they fail:
```rust
#[test]
fn test_password_characters_all() {
        let set = PasswordGeneratorCharSet::new(true, true, true, true, true);
        assert_eq!(set.lower, LOWER_CHARS, "set contains lower chars");
        assert_eq!(set.upper, UPPER_CHARS, "set contains upper chars");
        ...
    }
```

This way at least the error message will include a bit more detail: 
```
running 1 test
thread 'tool::generators::password::test::test_password_characters_all' panicked at crates/bitwarden/src/tool/generators/password.rs:201:9:
assertion `left == right` failed: set contains lower chars
  left: []
 right: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']

```

Reply:
I've found the unit test guidance for .Net to be absolutely great when it comes to language-agnostic advice. [Here's what it says](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests) about unit test names:

> The name of your test should consist of three parts:
>
>  * The name of the method being tested.
>  * The scenario under which it's being tested.
>  * The expected behavior when the scenario is invoked.
>
> ...Just by looking at the suite of unit tests, you should be able to infer the behavior of your code without even looking at the code itself. Additionally, when tests fail, you can see exactly which scenarios don't meet your expectations.

There's lots of ways to structure this data and everyone has their own opinion about the specifics, so I don't want to deep dive into exactly how to express this in rust code. My main goal is for the test methods to be more thoroughly described. There's one other benefit: using separate functions for each test isolates stateful operations, preventing them from affecting future tests.

> ...beyond splitting the tests into more functions with more descriptive names, the assert macros also have messages that they can print when they fail

This works only so long as your individual asserts reflect the higher-order goals of the test. Many tests require more than one assertion, which is why I'm focused on the function name.

Last but not least, rustdoc can [execute code examples as tests](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html). I'll leave figuring out if we'd like to leverage that for the future.